### PR TITLE
Quarkus NativeImageBuildStep fails with perm denied with docker rootless

### LINKED ISSUE #26890: Quarkus NativeImageBuildStep fails with perm denied with docker rootless
### Describe the bug

Building a minimal Quarkus project in native mode via docker rootless fails.

### Expected behavior

The build should not fail.

### Actual behavior

The build runs:
```
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildRunner] docker run --env LANG=C --rm --user 1000:1000 -v /home/rhousni/projects/quarkus-poc/target/to-csv-1.0-SNAPSHOT-native-image-source-jar:/project:z --name build-native-bENIH quay.io/quarkus/ubi-quarkus-native-image:22.1-java17 -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-Duser.language=en -J-Duser.country=US -J-Dfile.encoding=UTF-8 -H:-ParseOnce -J--add-exports=java.security.jgss/sun.security.krb5=ALL-UNNAMED -J--add-opens=java.base/java.text=ALL-UNNAMED -H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy\$BySpaceAndTime -H:+JNI -H:+AllowFoldMethods -J-Djava.awt.headless=true -H:FallbackThreshold=0 --link-at-build-time -H:+ReportExceptionStackTraces -H:-AddAllCharsets -H:EnableURLProtocols=http -H:NativeLinkerOption=-no-pie -H:-UseServiceLoaderFeature -H:+StackTrace to-csv-1.0-SNAPSHOT-runner -jar to-csv-1.0-SNAPSHOT-runner.jar
```

then fails with:
```
/gcc/x86_64-redhat-linux/8/crtend.o /usr/lib/gcc/x86_64-redhat-linux/8/../../../../lib64/crtn.o
/usr/bin/ld: cannot open output file /project/to-csv-1.0-SNAPSHOT-runner: Permission denied
collect2: error: ld returned 1 exit status
	at com.oracle.svm.hosted.image.NativeImageViaCC.handleLinkerFailure(NativeImageViaCC.java:505)
	at com.oracle.svm.hosted.image.NativeImageViaCC.write(NativeImageViaCC.java:452)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:695)
	at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:515)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:407)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:585)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:128)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus.main(NativeImageGeneratorRunner.java:615)
------------------------------------------------------------------------------------------------------------------------
                        1.5s (4.2% of total time) in 23 GCs | Peak RSS: 6.19GB | CPU load: 10.81
========================================================================================================================
Failed generating 'to-csv-1.0-SNAPSHOT-runner' after 35.7s.
Error: Image build request failed with exit status 1
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  42.163 s
[INFO] Finished at: 2022-07-22T19:55:56+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:2.10.3.Final:build (default) on project to-csv: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.deployment.pkg.steps.NativeImageBuildStep#build threw an exception: io.quarkus.deployment.pkg.steps.NativeImageBuildStep$ImageGenerationFailureException: Image generation failed. Exit code: 1
[ERROR] 	at io.quarkus.deployment.pkg.steps.NativeImageBuildStep.imageGenerationFailed(NativeImageBuildStep.java:400)
[ERROR] 	at io.quarkus.deployment.pkg.steps.NativeImageBuildStep.build(NativeImageBuildStep.java:240)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
[ERROR] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:944)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
[ERROR] 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:833)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

### How to Reproduce?

1. Make sure `docker` is installed in rootless mode => https://docs.docker.com/engine/security/rootless/
2. Create an empty Quarkus Maven project
3. Run `./mvnw clean package -P native`

### Output of `uname -a` or `ver`

```
Linux rhousni 5.15.0-41-generic #44~20.04.1-Ubuntu SMP Fri Jun 24 13:27:29 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

### Output of `java -version`

```
openjdk version "17" 2021-09-14 OpenJDK Runtime Environment (build 17+35-2724) OpenJDK 64-Bit Server VM (build 17+35-2724, mixed mode, sharing)
```

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

`2.10.3`

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

```
Apache Maven 3.8.6 (84538c9988a25aec085021c365c560670ad80f63) Maven home: /home/rhousni/.m2/wrapper/dists/apache-maven-3.8.6-bin/67568434/apache-maven-3.8.6 Java version: 17.0.1, vendor: Eclipse Adoptium, runtime: /home/rhousni/.jdks/temurin-17.0.1 Default locale: en_US, platform encoding: UTF-8 OS name: "linux", version: "5.15.0-41-generic", arch: "amd64", family: "unix"
```

### Additional information

Maybe linked to #10637 ?

### PR DESCRIPTION
**Pull request by reda-alaoui**
Fix #26890

### CODE REVIEW

**Review by zakkak — COMMENTED** [review]
As mentioned in #26890 this is related to https://github.com/moby/moby/issues/42134#issuecomment-796728856
Unfortunately though there is no proper solution there. There seem to be two options for this issue:

1. Run as root inside the container as this PR does (I wonder if that results in files been created inside the container to be owned by root on the host)
2. Adjust the permissions of the volumes being mounted on the host to match the actual user/group id of the user in the container.

The first approach is certainly simpler, but results in running the builder image container as root, which is not desirable in general.
The second approach is a bit more complex and requires careful handling to make sure permissions are set properly before and after the builder image runs.

I need to setup docker in rootless mode in order to toy around with this options, which I will probably not have the time to do in the following weeks, so if there is anyone else willing to experiment and review this PR in the meantime please do so.

**Comment by reda-alaoui** [comment]
Could someone at least trigger the build?

**Comment by geoand** [comment]
@zakkak what should we do with this one?

**Comment by geoand** [comment]
Sounds good

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildLocalContainerRunner.java [hunk 1]
```diff
@@ -2,37 +2,73 @@
...
 
+    private static boolean isDockerRootless(ContainerRuntimeUtil.ContainerRuntime containerRuntime) {
+        if (containerRuntime != ContainerRuntimeUtil.ContainerRuntime.DOCKER) {
+            return false;
+        }
+        String dockerHost = System.getenv("DOCKER_HOST");
+        // docker socket?
+        String socketUriPrefix = "unix://";
+        if (dockerHost == null || !dockerHost.startsWith(socketUriPrefix)) {
+            return false;
```
Comment (zakkak) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildLocalContainerRunner.java:54]: One can run docker rootless without setting `DOCKER_HOST`, see https://docs.docker.com/engine/security/rootless/#client

**Review by zakkak — CHANGES_REQUESTED** [review]
I managed to find some time and tried the PR with docker rootless and it works as expected, but requires setting the `DOCKER_HOST` env variable which is not necessary in order to use docker rootless see https://docs.docker.com/engine/security/rootless/#client.

Note: Unfortunately I will probably not be able to explore alternative options before the 19th of September.

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildLocalContainerRunner.java (hunk 1 above)
Comment (reda-alaoui) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildLocalContainerRunner.java]: Indeed. 

New proposition: call `docker context ls` if `DOCKER_HOST` is undefined.

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildLocalContainerRunner.java [hunk 2]
```diff
@@ -2,37 +2,112 @@
...
+        // DOCKER_HOST environment variable overrides the active context
+        String dockerHost = System.getenv("DOCKER_HOST");
+        if (dockerHost != null) {
+            return dockerHost;
+        }
+
+        OutputFilter outputFilter = new OutputFilter();
+        if (!ExecUtil.execWithTimeout(new File("."), outputFilter, Duration.ofMillis(3000),
+                "docker", "context", "ls", "--format",
+                "'{{- if .Current -}} {{- .DockerEndpoint -}} {{- end -}}'")) {
```
Comment (zakkak) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildLocalContainerRunner.java:90]:
```suggestion
                "{{- if .Current -}} {{- .DockerEndpoint -}} {{- end -}}")) {
```

With the `'` it returns `'''unix:///run/user/1000/docker.sock'` which fails the `startsWith("unix://")` check

**Review by zakkak — CHANGES_REQUESTED** [review]

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildLocalContainerRunner.java (hunk 2 above)
Comment (zakkak) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildLocalContainerRunner.java:90]: It's essentially like running:
```
$ docker context ls --format "'{{- if .Current -}} {{ .DockerEndpoint -}} {{- end -}}'"
''
'unix:///run/user/1000/docker.sock'
```

Comment (reda-alaoui) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildLocalContainerRunner.java]: Good catch

**Review by zakkak — APPROVED** [review]
Although I am not a big fan of running the container image as `root` I confirm that the PR fixes the issue and is good for merging till we figure out a better alternative.

Thank you for your contribution @reda-alaoui!

**Comment by reda-alaoui** [comment]
Are you waiting for something/someone to merge? :)

**Comment by geoand** [comment]
Can you please squash and rebase onto main?

**Comment by reda-alaoui** [comment]
> Can you please squash and rebase onto main?

Done :+1:

**Review by zakkak — APPROVED** [review]

**Comment by geoand** [comment]
I suppose we want this in `2.13`? If so, please add the `triage/backport` label.

**Comment by zakkak** [comment]
> I suppose we want this in `2.13`? If so, please add the `triage/backport` label.

I think it would indeed be **nice to have**.

**Pull request merged**